loader.py

```diff
@@ -0,0 +1,349 @@
+PIECES = {'P': 0, 'N': 1, 'B': 2, 'R': 3, 'Q': 4, 'K': 5,
+          'p': 6, 'n': 7, 'b': 8, 'r': 9, 'q': 10, 'k': 11}
+
+
+class FENGenerator:
+    def __init__(self, replay_queue: mp.Queue, pgn_path):
+
+        """
+        An iterable class that takes game saves and converts them into a chessboard position record in fen notation.
+        :param replay_queue: target queue for samples
+        :param pgn_paths: list of paths to data files
+        """
+
+        self.replay_queue = replay_queue
+        self.error_tolerance=2
+        self.current_path = (
+            0  # index of the path to the currently used data file in the pgn_paths list
+        )
+        self.current_line = (
+            []
+        )  # currently used line from the data file, representing one game
+        self.board = chess.Board()  # the chessboard on which the moves are made
+        if isinstance(pgn_path,str):
+            self.pgn_paths = [pgn_path]
+        else:
+            self.pgn_paths = pgn_path
+        # pgn_file - currently used data file
+        self.pgn_file = None
+        self.get_new_file(0)
+
+    def closeFile(self):
+        if self.pgn_file != None:
+            self.pgn_file.close()
+
+    def get_new_file(self, index):
+        if self.pgn_file != None:
+            self.pgn_file.close()
+        if ".bz2" in self.pgn_paths[index]:
+            self.pgn_file = bz2.open(self.pgn_paths[index])
+        elif "https://" in self.pgn_paths[index]:
+            self.pgn_file = urlopen(self.pgn_paths[index])
+        else:
+            self.pgn_file = open(self.pgn_paths[index])
+
+    def get_line(self):
+        """
+        get a new line with a new game from the currently used data file into the variable current_line.
+        If all lines from the file have already been read,
+        set the next file from the list pgn_paths to the pgn_file and read the first line
+        :return: None
+        """
+        line = self.pgn_file.readline()
+        if isinstance(line,bytes):
+            line = line.decode("utf-8")
+        if line == "":
+            self.current_path += 1
+            if self.current_path >= len(self.pgn_paths):
+                self.current_path = 0
+            self.get_new_file(self.current_path)
+            line = self.pgn_file.readline()
+        self.current_line = line[:-1].split()
+
+    def get_position(self,n_error=0):
+        """
+        It makes the next move on the list and reads the positions on the board
+        if it fails, I load another game from the data file and repeat the process
+        :return: String of chessboard positions in FEN notation
+        """
+        if len(self.current_line) < 1:
+            self.get_line()
+            self.board.reset()
+        try:
+            self.board.push_san(self.current_line[0])
+            self.current_line.pop(0)
+            return self.board.fen()
+        except Exception as ex:

```

Comment:
I am against that mechanism - it will always allow any two errors, regardless of their type, and we don't control what we're allowing. I'm curious, what is the story behind that?

Reply:
I made this fix quickly, the point is that if there is, for example, a bug in the dataset, the code should not crash, but skip the game with the error. I will think about it yet.

Reply:
I doubt we need some error tolerance, especially in this form, because currently when error is raised one doesn't know what to do to prevent error next time. In my opinion we should silently ignore all errors or remove games which contain errors (this can be done for example during preprocessing).

And according to the docs this method can raise only `ValueError`, so it can be used instead of `Exception`.